protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/IntentTypeIdResolver.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.protocol.jackson.record;
+
+import com.fasterxml.jackson.annotation.JsonTypeInfo.Id;
+import com.fasterxml.jackson.databind.DatabindContext;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.jsontype.impl.TypeIdResolverBase;
+import io.camunda.zeebe.protocol.record.ValueType;
+import io.camunda.zeebe.protocol.record.intent.Intent;
+
+/**
+ * A {@link com.fasterxml.jackson.databind.jsontype.TypeIdResolver} that maps a serialized {@link
+ * Intent} value to a concrete implementation enum, e.g. {@link
+ * io.camunda.zeebe.protocol.record.intent.VariableIntent}, based on the value type of the record.
+ */
+final class IntentTypeIdResolver extends TypeIdResolverBase {
+  @Override
+  public String idFromValue(final Object value) {
+    return ((ValueType) value).name();
+  }
+
+  @Override
+  public String idFromValueAndType(final Object value, final Class<?> suggestedType) {
+    return idFromValue(value);
+  }
+
+  @Override
+  public Id getMechanism() {
+    return Id.CUSTOM;
+  }
+
+  @Override
+  public JavaType typeFromId(final DatabindContext context, final String id) {
+    final var valueType = ValueType.valueOf(id);
+    final var typeFactory = context.getTypeFactory();
+    return typeFactory.constructType(mapValueTypeToIntentClass(valueType));
+  }
+
+  private Class<? extends Intent> mapValueTypeToIntentClass(final ValueType valueType) {
+    final ValueTypeInfo<?> typeInfo = ValueTypes.getTypeInfoOrNull(valueType);
+    if (typeInfo == null) {
+      return Intent.UNKNOWN.getClass();
```

Comment:
:question: This keeps it aligned with how we handle records internally, but I don't know if that's a good idea or unexpected from an external user PoV?

Reply:
❓ What is the alternative? Throwing an exception or using null? I think what you've provided is the best out of those.

Reply:
:thought_balloon: It's a deserialization failure imo, so I would let it throw a `JsonProcessingException` error as an alternative. I'm actually not sure why we also have the `Unknown` intent in the first place :smile: 

Let's say this happens at runtime. How is your code supposed to react to this? We're now leaving it up to the user to check if it's the expected intention. Now let's say the user has some code like:

```java
final var record = deserialize(myJson);
if (record.getIntent() == Intent.UNKNOWN) {
  explodeOrDoSomethingWithIt();
}
```

How would they introspect what the original intent was though? I guess via the JSON...that's probably fine for most users I think.

Reply:
I see your point, but throwing it in the `mapValueTypeToIntentClass` scope does not give access to the json directly right? Or does jackson decorate the exception with useful data if you throw `JsonProcessingException` inside that method? If that's the case, then let's fail early. Otherwise, let's fail late and allow the user to discover the context

Reply:
Let's leave it as is then.